Rebuild js-support page for Docusaurus v3

didoesdigital opened the pull request:
This PR builds on the prototype in https://github.com/yeslogic/prince-docs/pull/44

The fundamental change here is assuming that JavaScript property "type" information exists on `std.json` and that `std.json` is produced by Prince. This lets us update the React code to use that info from the new static `_std-data.json` file (that combines `std.json` with `std-annotated.json`) instead of asking the browser's `window` object for this info at run time.

This approach lets us avoid using `BrowserOnly` by never accessing the `window` object so we can then statically generate the file without client-side code. That was discussed in https://github.com/yeslogic/prince-docs/pull/40. This also simplifies some things, making it easier for the React code to just render what it's been given. 

Notable changes on this PR:

- added `std_v3.json` from https://github.com/yeslogic/prince-docs/pull/44#issuecomment-3247872784 as `std.json` to this PR
- performed a 1-time conversion of `std-annotated.js` to JSON as `std-annotated.json` with properties in double quotes, etc.
- removed redundant `type` information for properties from `std-annotated.json` that is now generated (but kept `type` info for `arguments` on `std-annotated.json` because we won't be able to generate those)

Later on, we could potentially add @adrianwong's script to generate `std.json` to Prince and update the docs' repo's `Makefile` build process to run that Prince command as part of the docs site build to reproduce that `std.json` over time.

Some visuals of what the prototype currently looks like:

<img width="1277" height="711" alt="image" src="https://github.com/user-attachments/assets/eb8f6122-88bf-438e-af02-cf1e3d71e32a" />

<img width="1276" height="708" alt="image" src="https://github.com/user-attachments/assets/d60fff00-e387-4666-b230-bb07e82d7ef4" />

<img width="1276" height="673" alt="image" src="https://github.com/user-attachments/assets/b720b929-48a6-431e-982a-99877b12e9c4" />

<img width="1272" height="706" alt="image" src="https://github.com/user-attachments/assets/a4d1d174-2f4b-4a95-9a5f-b0c0d57131fa" />

<img width="1277" height="713" alt="image" src="https://github.com/user-attachments/assets/653f189a-502a-4530-bc14-9a9d87da9d88" />

<img width="1277" height="716" alt="image" src="https://github.com/user-attachments/assets/60611fe1-9d12-41c1-9bfe-b327b607f23d" />

<img width="1276" height="715" alt="image" src="https://github.com/user-attachments/assets/d0607cbe-c220-43fe-9024-fda910f8764d" />

Other notes:

- looks like `jq` is already [available in the Ubuntu 24 image in the GitHub runner](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#tools) so there should be no need to install it in the GitHub workflows

didoesdigital committed:
Add std-annotated.js quotes before JSON conversion

Use `git show THIS_COMMIT --color-words` to more easily see that this
commit mostly only adds double quotes to the stdAnnotated object's keys.
Used JSON.stringify(stdAnnotated, null, 4), which technically also
tweaked some whitespace, line breaks, and wrapping of curly braces.

didoesdigital committed:
Convert std-annotated.js to std-annotated.json

didoesdigital committed:
Move std-annotated.json into src

didoesdigital committed:
Replace std.js with std.json with __princetype__

This file was generated by @adrianwong in
https://github.com/yeslogic/prince-docs/pull/44#issuecomment-3204154110

didoesdigital committed:
Remove types from std-annotated.json

We will instead be relying on the __princetype__ field in std.json

didoesdigital committed:
Update std.json with Adrian's std_v3.json changes

https://github.com/yeslogic/prince-docs/pull/44#issuecomment-3247872784

didoesdigital committed:
Add top-level window "desc" to std-annotated.json

didoesdigital committed:
Fix typo of "desc" in std-annotated.json

didoesdigital committed:
Update build of JS Support page

- Add Makefile for js-support page to combine std.json and
  std-annotated.json into _std-data.json using jq
- Remove BrowserOnly and client-side behaviour in favour of using type
  information in __princetype__ key from _std-data.json
- Move "See the property " from CSS pseudo elements into JSX/HTML
- Move js-support.md content into JS to add <span className="dep">

didoesdigital committed:
Remove unused style file

didoesdigital committed:
Amend annotations in std-annotated.json

Remove manual "type" annotations in std-annotated.json and adjust other
fields like "returns" to use "desc" instead because they no longer have
their types documented as "function"

https://github.com/yeslogic/prince-docs/pull/44#issuecomment-3241673483

didoesdigital committed:
Restore HTMLDocument.write in js-support

didoesdigital committed:
Remove "innerHTML" from annotations

No longer in std.json

didoesdigital committed:
Remove String.CodePointAt from annotated

There's already String.prototype.codePointAt (lowercase "c")

didoesdigital committed:
Remove unsupported DOMTokenList from annotated

didoesdigital committed:
Convert comments to proper JSDocs docstrings

This will let the props show up on hover in editors

didoesdigital committed:
Document auto-generated file

didoesdigital committed:
Extract openDetails function

didoesdigital committed:
Extract useHashDetails hook

didoesdigital committed:
Extract strings as arguments in new toggleAllItems

didoesdigital committed:
Extract toggleAllItems to its own file

didoesdigital committed:
Add guard against invalid strings used as id

didoesdigital committed:
Adjust id for #.window on js-support page

didoesdigital committed:
Reuse useHashDetails, toggleAllItems on js-support

didoesdigital committed:
Adjust toggleAllItems for js-support behaviour

didoesdigital committed:
Remove unused scripts, imports

script and shiftWindow.js are now replaced by toggleAllItems.js,
useHashDetails.js, openDetails.js

didoesdigital committed:
Give global window item grey background

didoesdigital committed:
Extract the JavascriptSupportItem sub-components

didoesdigital committed:
Document the JavaScript support page and jq

didoesdigital committed:
Add js-support page steps to GitHub workflows

didoesdigital committed:
Remove unused import

didoesdigital commented:
@csant would you please check out this branch, [install `jq`](https://jqlang.org/download/), run `make && yarn run build && yarn run serve`, inspect it the JavaScript Support page in the browser and confirm that it is complete and behaves correctly?

I expect that when you run `make` you should see no git diffs on the `_std-data.json` file, but please let me know if it builds it differently. You could even test deleting `_std-data.json` and running `make` again to confirm that it's building correctly.

On the JavaScript Support page, you should be able to see:

- The intro blurb includes a grey deprecated line-through
- Toggling all items, linking to and scrolling to sections should work for all scenarios, just like [the behaviour on the CSS page](https://github.com/yeslogic/prince-docs/pull/43)
- The CSS page interactions should also still work, but now with extracted out shared logic between the two 
- `NaN` and `prototype` and all the weird items should appear fine.
- All the special fields like `url`, `dep`, `ext` should do what they need to, linking to the CSS page, greying out deprecated entries etc.
- The JSDocs docstrings in `/** */` should appear when you hover over props (see image below)

<img width="986" height="260" alt="image" src="https://github.com/user-attachments/assets/b8a02333-180b-4c63-969a-2c3316edcf22" />

didoesdigital commented:
Some more details I noticed in `std-annotated.json`:

- `PDF.openAction` and `PDF.profile` seem to be `object` according to `std.json` and I've amended them like the other `PDF` entries to use `desc` instead of `arguments` fields, but now the example doesn't look quite right: e.g. `PDF.openAction('print')`. @csant after we've merged this PR, could you please review what the examples should be?
- FYI, I removed `HTMLDocument.prototype.innerHTML` from `std-annotations.json` because it's no longer in `std.json`
- FYI, I removed `String.CodePointAt` (note the capital "C"), because `std.json` has `String.prototype.codePointAt` (lowercase "c") on `prototype` instead
- I removed these from `DOMTokenList.prototype.` because they're not supported:
    ```
      "value": {},
      "supports": {},
      "entries": {},
      "forEach": {},
      "keys": {},
      "values": {}
    ```
- `Error.prototype.stack` is missing from `std.json` but I was less sure about removing that one from `std-annotated.json` @adrianwong can you confirm if I should? <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/stack>

csant commented:
> @csant would you please check out this branch, [install `jq`](https://jqlang.org/download/), run `make && yarn run build && yarn run serve`, inspect it the JavaScript Support page in the browser and confirm that it is complete and behaves correctly?

That worked smoothly, and all seems to be looking fine.

> I expect that when you run `make` you should see no git diffs on the `_std-data.json` file, but please let me know if it builds it differently. You could even test deleting `_std-data.json` and running `make` again to confirm that it's building correctly.

`_std-data.json` is not a tracked - should it be?

[...]

> * The JSDocs docstrings in `/** */` should appear when you hover over props (see image below)

Not really sure what you mean by that.

csant commented:
> Later on, we could potentially add @adrianwong's script to generate `std.json` to Prince and update the docs' repo's `Makefile` build process to run that Prince command as part of the docs site build to reproduce that `std.json` over time.

It would be very good to have the script to be able to generate the `std.json` when needed (basically, each time new props are added). This requires however also manual intervention on `std-annotated.json`, so it cannot be part of an automated process.

csant commented:
> `_std-data.json` is not a tracked - should it be?

Ignore this comment - all works fine, and the file is tracked. I was looking at `_std-data.json` in the output folder... 🙄

csant commented:
@didoesdigital I noticed that links in `desc` are correctly parsed as HTML - except for the desc in `convertToFile()`.
A `desc` inside of `arguments` seems to be parsed as a text string, rather than HTML.

csant commented:
@didoesdigital @adrianwong I notice that `princePdfColorOptions` is missing from `std.json`.

adrianwong commented:
> * `PDF.openAction` and `PDF.profile` seem to be `object` according to `std.json` and I've amended them like the other `PDF` entries to use `desc` instead of `arguments` fields, but now the example doesn't look quite right: e.g. `PDF.openAction('print')`. @csant after we've merged this PR, could you please review what the examples should be?

Can confirm that `PDF.openAction` and `PDF.profile` are not functions. Both should in fact be `string`. I've noticed that these two (and a number of other properties e.g. `userPassword`) return `null` if they haven't been set, which unfortunately classifies them as `object`.

If you'd like, I can trawl through the source code and manually fix up `std.json` for these cases as best as I can.

-----

> * `Error.prototype.stack` is missing from `std.json` but I was less sure about removing that one from `std-annotated.json` @adrianwong can you confirm if I should?

Seems like it doesn't show up because of the way the `Error` object is coded, whereby the `stack` property only gets added at the point where an error is thrown.

I can manually add this one, though I'm also fine with keeping this property out of the documentation. From memory it was added for internal debugging purposes, and it doesn't appear to be part of the standard (of the "big three", only Firefox has this property on the `Error` prototype).

adrianwong commented:
> @didoesdigital @adrianwong I notice that `princePdfColorOptions` is missing from `std.json`.

Hmmm. What I'm seeing in the source code is that `princePdfColorOptions` may have never been a property of `CSSStyleDeclaration`. Rather, from the way it (and other properties using a similar pattern) have been coded, it is accessed via `PDF.colorOptions` instead.

csant commented:
> Hmmm. What I'm seeing in the source code is that `princePdfColorOptions` may have never been a property of `CSSStyleDeclaration`. Rather, from the way it (and other properties using a similar pattern) have been coded, it is accessed via `PDF.colorOptions` instead.

@adrianwong oh, true. What a mess - some -prince-pdf-* CSS properties are part of the `CSSStyleDeclaration`, some aren't. What is more, there is no `color-options` (or even `pdf-color-options`) alias to the property `-prince-pdf-color-options`. At some point it might have ended up there because I added it manually, (wrongly) assuming that (all) CSS properties were also accessible through `CSSStyleDeclaration`. Shouldn't they be?

This however would speak against adding things manually to `std.json`, or fixing them manually - any new entry would need to be added (or reviewed) by you, to avoid wrong assumptions.

Coming back to your further comments above, about fixing some output in `std.json` - is there no way to detect these exceptions in the script generating it? Or are there way too many exceptions?  Because at some point it would be good to have that file generated by Prince (and only add things manually to `std-annotated.json`), so we can always be sure that the documentation truly reflects what Prince supports.

adrianwong commented:
> What a mess - some -prince-pdf-* CSS properties are part of the `CSSStyleDeclaration`, some aren't. What is more, there is no `color-options` (or even `pdf-color-options`) alias to the property `-prince-pdf-color-options`. At some point it might have ended up there because I added it manually, (wrongly) assuming that (all) CSS properties were also accessible through `CSSStyleDeclaration`. Shouldn't they be?

Perhaps @mikeday can comment on this, as I am uncertain as to why this is the case.

> Coming back to your further comments above, about fixing some output in `std.json` - is there no way to detect these exceptions in the script generating it? Or are there way too many exceptions? Because at some point it would be good to have that file generated by Prince (and only add things manually to `std-annotated.json`), so we can always be sure that the documentation truly reflects what Prince supports.

Good point. In terms of _properties_, the only ones that I manually added per [this comment](https://github.com/yeslogic/prince-docs/pull/44#issuecomment-3247872784) were excluded by the script as they were aliases; this was done for backward compatibility with current documentation. I.e., the script will pick up these properties _once_ (and hence will reflect the current state of Prince), but they may not necessarily be on the object we want (for documentation purposes, that is).

In [this comment](https://github.com/yeslogic/prince-docs/pull/45#issuecomment-3276983519) I proposed manually adding `Error.prototype.stack`; in light of your comment perhaps this isn't such a good idea.

In terms of property _types_, there are two issues:

* The first is properties that throw an error when accessed. These are numerous (470 in `_v3.json`), but are caught by the script and assigned a type of `null`. There might be ways to fix up some, but likely not all.
* The second ([mentioned here](https://github.com/yeslogic/prince-docs/pull/45#issuecomment-3276983519)) is that our API sometimes returns a `null` to indicate that a property hasn't been set, but its set type is not an object. It's trivial to update the script to check for these, but it will still require manual fixes.

-----

tl;dr Unless we employ the pattern of adding a property late à la `Error.prototype.stack`, the script should pick up any new properties we add. However, their listed types may not necessarily be correct.

didoesdigital committed:
Treat arguments.desc content as HTML not string

didoesdigital committed:
Fix up component JSDocs for props

didoesdigital commented:
> @didoesdigital I noticed that links in `desc` are correctly parsed as HTML - except for the desc in `convertToFile()`. A `desc` inside of `arguments` seems to be parsed as a text string, rather than HTML.

I've updated it to treat `arguments.desc` as HTML, fixing the link in `convertToFile()` https://github.com/yeslogic/prince-docs/pull/45/commits/6df58e7e977ceca3c83d9516ed9d02dfe7d59163

<img width="1778" height="644" alt="convertToFileHTML" src="https://github.com/user-attachments/assets/e0415ac9-5da7-42bb-8539-18e2c1521e61" />

didoesdigital commented:
@csant Regarding my comment "The JSDocs docstrings in `/** */` should appear when you hover over props", I meant that for all the components related to `JavascriptSupportItem` I have added comments using [JSDoc](https://jsdoc.app/) comment syntax so that when you hover over a component property in an editor like Visual Studio Code or similar, it should show those comments. It was helpful to me while building this PR to document the React component types and their expected values and be able to see those examples where the components are used. I fixed up some of them in https://github.com/yeslogic/prince-docs/pull/45/commits/8c15f3853fdd131455b3e05a46027015d99c25aa

didoesdigital committed:
Fix PDF.openAction and PDF.profile examples

didoesdigital committed:
Add __princetype__ override in std-annotated.json

didoesdigital commented:
I've updated the `PDF.openAction` and `PDF.profile` examples in https://github.com/yeslogic/prince-docs/pull/45/commits/acb79c51e1921d82f1568d69efbbb081c334dc22 to use strings

<img width="1948" height="538" alt="PDF.openAction = 'print'" src="https://github.com/user-attachments/assets/da98e00e-7b0c-45f9-ad3c-7b2d16c967ac" />

<img width="1930" height="494" alt="PDF.profile = 'PDF/X-3:2003'" src="https://github.com/user-attachments/assets/bc2456af-7b23-4238-80a0-34085e397788" />

didoesdigital commented:
The way the `std.json` and `std-annotated.json` files are merged into `_std-data.json` in the `Makefile` process lets the annotated version override fields in the `std.json` version, so I've added a commit demonstrating how we might manually patch things in `std-annotated.json` and keep `std.json` automated and repeatable https://github.com/yeslogic/prince-docs/pull/45/commits/ac6c13e54a682e4b03b043a915230c075a91e40d

It might be slightly misleading that `std-annotated.json` has `__princetype__` in it when it's not info officially reported by Prince but maybe it's a pragmatic enough compromise?

What do you think of using this override approach for  @csant @adrianwong ?

didoesdigital committed:
Remove "princePdfColorOptions" JS docs entry

See: https://github.com/yeslogic/prince-docs/pull/45

didoesdigital commented:
I've removed `"princePdfColorOptions"` from `std-annotated.json` in https://github.com/yeslogic/prince-docs/pull/45/commits/62950fc4f111b1ae195baf86a1f96e90d55a2e9e for now. I think we can address the other questions around the `-prince-pdf-*` CSS properties separately from this PR.

csant commented:
> It might be slightly misleading that `std-annotated.json` has `__princetype__` in it when it's not info officially reported by Prince but maybe it's a pragmatic enough compromise?

Does it actually matter?  `std-annotated.json` is not meant to be used as documentation, the produced HTML is.

We shouldn't worry about people looking at `std-annotated.json` and seeing some "inner processes", since they can equally well inspect `std.json` on Github, and see `__princetype__`.

Might the override approach allow us to "patch" all issues that would need us to manually fix `std.json`?

adrianwong commented:
> Might the override approach allow us to "patch" all issues that would need us to manually fix `std.json`?

As it stands, I think so.

> We shouldn't worry about people looking at `std-annotated.json` and seeing some "inner processes", since they can equally well inspect `std.json` on Github, and see `__princetype__`.

I'll defer to you on this @csant. If you're okay with @didoesdigital's approach then so am I.

didoesdigital commented:
Ok! I think that means we can merge this PR now? @csant?
